server/aesdsocket.c

```diff
@@ -23,217 +23,313 @@
         close(sockfd);
         perror("Error binding \n");
-        return -1;
+        //return -1;
     }
 
     //daemon creation if -d is present
     //daemon kill handling is pending
-    if(argc > 1)
-    {
-        if(!strcmp(argv[1],"-d"))
-        {
+    if (argc > 1) {
+        if (!strcmp(argv[1], "-d")) {
             //create_deamon();
             pid_t pid = fork();
 
-    		if (pid < 0) 
-    		{
-    			printf("error:pid<0");
-        		exit (-1);
-    		}
-    		else if (pid == 0)
-    		{
-    			if (setsid() < 0) 
-    			{
-        			perror("Cannot set SID");
-        			exit (-1);
-        		}
-    		
-    			if(chdir("/") == -1)
-    			{
-    				perror("error changing directory");
-    				exit (-1);
-    			} 
-    			open("/dev/null", O_RDWR);
-    			dup(0);
-    			dup(0);
-    			printf("Daemon created successfully\n");
-    			//exit (0);
-    		}
-            else {
-                exit (0);
+            if (pid < 0) {
+                printf("error:pid<0");
+                exit(-1);
+            } else if (pid == 0) {
+                if (setsid() < 0) {
+                    perror("Cannot set SID");
+                    exit(-1);
+                }
+
+                if (chdir("/") == -1) {
+                    perror("error changing directory");
+                    exit(-1);
+                }
+                open("/dev/null", O_RDWR);
+                dup(0);
+                dup(0);
+                printf("Daemon created successfully\n");
+                //exit (0);
+            } else {
+                exit(0);
             }
         }
     }
 
-    printf("Listening\n");
+    printf("opening /var/tmp/aesdsocketdata\n");
+    file_fd = open("/var/tmp/aesdsocketdata", O_RDWR | O_CREAT | O_TRUNC,
+                   S_IRUSR | S_IWUSR | S_IRGRP | S_IWGRP | S_IROTH | S_IWOTH);
+
+    if (file_fd == -1) {
+        perror("error opening the file");
+    }
+    printf("file successfully opened\n");
+    
+    //mutex init
+    pthread_mutex_init(&lock, NULL);
+
+    TAILQ_INIT(&head);
+
     int sockls;
+    bool alarm_flag = false;
+
+    printf("Listening\n");
     sockls = listen(sockfd, 5);
-    printf("Listening successful\n");
+    printf("sockls value is %d\n", sockls);
+    printf("sockfd value is %d\n", sockfd);
 
-    if (sockls == -1)
-    {
+    if (sockls == -1) {
         perror("Error Listening \n");
-        return -1;
+        //return -1;
     }
+    printf("Listening successful\n");
 
-    printf("opening /var/tmp/aesdsocketdata\n");
-    file_fd = open("/var/tmp/aesdsocketdata",O_RDWR | O_CREAT | O_TRUNC, S_IRUSR | S_IWUSR | S_IRGRP |S_IWGRP | S_IROTH | S_IWOTH);
-    
-    if (file_fd == -1)
-    {
-        perror("error opening the file");
-    }
-    printf("file successfully opened\n");
-    int file_size = 0;
-    
-    char receive_buffer[BUFFER_SIZE];
-    
-    while(1)
-    {
+    while (1) {
+        thread_data_t *data = (thread_data_t *) malloc(sizeof(thread_data_t));
```

Comment:
@TejaswiniLakshminarayan what happens if the `TERM` signal occurs between tis line and line 310?  Where will the free of the data malloc'd here occur?